core/chains/solana/soltxm/cache.go

```diff
@@ -0,0 +1,120 @@
+package soltxm
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/gagliardetto/solana-go"
+	solanaClient "github.com/smartcontractkit/chainlink-solana/pkg/solana/client"
+	"golang.org/x/exp/maps"
+)
+
+type TxCache struct {
+	chainID string
+	cache   map[solana.Signature]context.CancelFunc
+	lock    sync.RWMutex
+}
+
+func NewTxCache(id string) *TxCache {
+	return &TxCache{
+		chainID: id,
+		cache:   map[solana.Signature]context.CancelFunc{},
+	}
+}
+
+func (c *TxCache) Insert(sig solana.Signature, cancel context.CancelFunc) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.cache[sig] != nil {
+		return errors.New("signature already exists")
+	}
+	c.cache[sig] = cancel
+	return nil
+}
+
+// Success - tx included in block and confirmed
+func (c *TxCache) Success(sig solana.Signature) {
+	promSolTxmSuccessfulTxs.WithLabelValues(c.chainID).Add(1)
+	c.cancel(sig)
+	return
+}
+
+// Revert - tx included in block but failed execution
+func (c *TxCache) Revert(sig solana.Signature) {
+	promSolTxmRevertedTxs.WithLabelValues(c.chainID).Add(1)
+	c.cancel(sig)
+	return
+}
+
+// Failed - tx failed sending to chain or failed simulation
+func (c *TxCache) Failed(sig solana.Signature) {
+	promSolTxmFailedTxs.WithLabelValues(c.chainID).Add(1)
+	c.cancel(sig)
+	return
+}
+
+// Cancel - tx retry timed out, was not picked up by the network and confirmed in time
+func (c *TxCache) Cancel(sig solana.Signature) {
+	promSolTxmTimedOutTxs.WithLabelValues(c.chainID).Add(1)
+	c.cancel(sig)
+	return
+}
+
+func (c *TxCache) cancel(sig solana.Signature) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// already cancelled
+	if c.cache[sig] == nil {
+		return
+	}
+
+	c.cache[sig]() // cancel context
+	delete(c.cache, sig)
+	return
+}
+
+func (c *TxCache) List() []solana.Signature {
+	c.lock.RLock()
+	sigs := maps.Keys(c.cache)
+	c.lock.RUnlock()
+
+	promSolTxmInflightTxs.WithLabelValues(c.chainID).Set(float64(len(sigs)))
+	return sigs
+}
+
+type ValidClient struct {
+	tc     func() (solanaClient.ReaderWriter, error)
+	client solanaClient.ReaderWriter
+	lock   sync.Mutex
+}
+
+func NewValidClient(tc func() (solanaClient.ReaderWriter, error)) *ValidClient {
+	return &ValidClient{
+		tc: tc,
+	}
+}
+
+// Get a new client if it doesnt already exist
+func (vc *ValidClient) Get() (solanaClient.ReaderWriter, error) {
```

Comment:
This is a specific lazy loading component that I feel can be completely generalized, especially with Go1.18 generics. Here is [an example](https://github.com/hymkor/go-lazy) we could potentially use/include, but also the underlying impl should use [sync.Once](https://pkg.go.dev/sync#Once) as it's optimized for these kinds of checks.

Reply:
👍🏼 gave the generic function and using `sync.Once` a go. let me know what you think (and also where it should go since it's not solana specific)

Reply:
Oh I think you were right about it being in conflict with `Clear()`... might be simpler without the once. I should have looked closer. Generic is good though :+1: - not sure where yet

Reply:
Maybe just give it a separate file for now, w/o solana imports. Then it can be relocated cleanly in the future.
And /nit, what about `Reset()` instead of `Clear()`? Has some precedent in std libs.